metadata-ingestion/src/datahub/ingestion/source/sql/two_tier_sql_source.py

```diff
@@ -0,0 +1,95 @@
+import typing
+from typing import Any, Dict
+
+import pydantic
+from pydantic.fields import Field
+from sqlalchemy import create_engine, inspect
+from sqlalchemy.engine.reflection import Inspector
+
+from datahub.configuration.common import AllowDenyPattern
+from datahub.emitter.mcp_builder import PlatformKey
+from datahub.ingestion.api.workunit import MetadataWorkUnit
+from datahub.ingestion.source.sql.sql_common import (
+    BasicSQLAlchemyConfig,
+    SQLAlchemySource,
+    logger,
+    make_sqlalchemy_uri,
+)
+
+
+class TwoTierSQLAlchemyConfig(BasicSQLAlchemyConfig):
+
+    database_pattern: AllowDenyPattern = Field(
+        default=AllowDenyPattern.allow_all(),
+        description="Regex patterns for databases to filter in ingestion.",
+    )
+    schema_pattern: AllowDenyPattern = Field(
+        default=AllowDenyPattern.allow_all(),
+        description="Deprecated in favour of database_pattern. Regex patterns for schemas to filter in ingestion. "
+        "Specify regex to only match the schema name. e.g. to match all tables in schema analytics, "
+        "use the regex 'analytics'",
+    )
+
+    @pydantic.root_validator()
+    def ensure_profiling_pattern_is_passed_to_profiling(
+        cls, values: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        schema_pattern = values.get("schema_pattern")
+        if schema_pattern is not None:
+            values["database_pattern"] = schema_pattern
+        return values
+
+    def get_sql_alchemy_url(
+        self,
+        uri_opts: typing.Optional[typing.Dict[str, typing.Any]] = None,
+        current_db: typing.Optional[str] = None,
+    ) -> str:
+        return self.sqlalchemy_uri or make_sqlalchemy_uri(
+            self.scheme,  # type: ignore
+            self.username,
+            self.password.get_secret_value() if self.password else None,
+            self.host_port,  # type: ignore
+            current_db if current_db else self.database,
+            uri_opts=uri_opts,
+        )
+
+
+class TwoTierSQLAlchemySource(SQLAlchemySource):
+    def __init__(self, config, ctx, platform):
+        super().__init__(config, ctx, platform)
+        self.current_database = None
+        self.config: TwoTierSQLAlchemyConfig = config
+
+    def get_table_container_key(self, db_name: str, schema: str) -> PlatformKey:
+        return self.gen_database_key(db_name)
+
+    def get_allowed_schemas(
```

Comment:
can you add a comment on why it is called allowed schema? I think without knowing why it is called as it is there is a risk somebody will refactor it